### ISSUE TITLE
Declare more TS typings

### ISSUE THREAD

**Issue by vidartf** [jupyter-widgets/ipywidgets#1672]
The current typescript code has a lot of anys (both explicit and implicit). These should be more explicitly typed, at least on the exported functions. A more explicit typing, maybe together with strict null checks might also reveal some bugs.

- [x] no explicit `any` typings unless absolutely necessary
- [x] no implicit `any` typings unless necessary
- [x] turn on strict null checking

**Comment by jasongrout** [comment]
Sounds like an excellent sprint day project.

**Comment by vidartf** [comment]
Yes, absolutely. If anybody wants to do it, that would be perfect, if not, I'll try to squeeze it in 😉

**Comment by cnishina** [comment]
Currently working on this. 😄

**Comment by cnishina** [comment]
Here is the examples that needs review. https://github.com/jupyter-widgets/ipywidgets/pull/1733

**Comment by vidartf** [comment]
@cnishina I did some work previously that stranded a bit, see #1734. Always meant to come back to it, but you know how it goes. If you are able to use that as a starting point to take further that would be awesome!

**Comment by cnishina** [comment]
@vidartf ☹️ I'm having some weird tsc errors that I did not get a few days ago. So I did an npm install in the main directory. I get the following error:  
```
> @jupyter-widgets/base@1.0.2 build:src /path/to/ipywidgets/packages/base
> tsc --project src

src/utils.ts(14,8): error TS2307: Cannot find module 'base64-js'.
```

That's weird, so I did an `npm install -D @types/base64-js` to add the typings from DefinitelyTyped. Then I get another error: 

```
> @jupyter-widgets/controls@1.0.2 build:src /path/to/ipywidgets/packages/controls
> tsc --project src

src/phosphor/accordion.ts(10,8): error TS2307: Cannot find module '@phosphor/signaling'.
src/phosphor/currentselection.ts(14,8): error TS2307: Cannot find module '@phosphor/signaling'.
src/phosphor/tabpanel.ts(14,8): error TS2307: Cannot find module '@phosphor/signaling'.
```

Super weird. I did a brand new clone, install and came up with the same sort of errors. Is there something that I'm missing in the developer setup?

**Comment by vidartf** [comment]
It seems a lot of this i being fixed in #1738.

**Comment by jasongrout** [comment]
Yes, it's weird that somehow these problems are just now being noticed.

**Comment by cnishina** [comment]
OOoOoOh...  going to check it out.

**Comment by cnishina** [comment]
Back on track but now I have new problems. I can add some explicit typings in so there are no implicit any's; however, some of these issues might require some additional work. For example...looking at `ipywidgets/packages/controls/widget_box.ts`

- `BoxView` extends type `DOMWidgetView`
- `DOMWidgetView` has property `pWidget`. `pWidget` is type `Widget`
- in `BoxView`, the property `pWidget` is being set to `JupyterPhosphorPanelWidget`
- `JupyterPhosphorPanelWidget` extends type `Panel` and is cannot be of type `Widget`

🤔

**Comment by cnishina** [comment]
I also started to make changes to use fat arrows and `let` over `var`; however, this project appears to be set up as ES5 based on the tsconfig. In hind sight I probably should have clarified this first. Jupyter widgets is using ES5?

**Comment by jasongrout** [comment]
> JupyterPhosphorPanelWidget extends type Panel and is cannot be of type Widget

Panel is a subclass of Widget: https://github.com/phosphorjs/phosphor/blob/23b9d075ebc5b73ab148b6ebfc20af97f85714c4/packages/widgets/src/panel.ts#L28

**Comment by jasongrout** [comment]
> Jupyter widgets is using ES5?

Jupyter widgets is using Typescript. Typescript is configured to compile arrow functions and let down to ES5 for backwards compatibility.

I was hoping to go through and mak a pass changing var to let, using arrow functions, etc. If you want to do it, go for it! (perhaps on a separate PR to make reviewing easier?)

**Comment by cnishina** [comment]
I've been replacing `that = this` and using fat arrows, and `let` over `var`. I just realized my previous comment about ES5 is moot because the tsconfig is transpiling to ES5 but TypeScript does not care.

**Comment by blink1073** [comment]
Yep, we have been treating TypeScript as ES6 + type information, not really using the experimental features like decorators.

**Comment by jasongrout** [comment]
The other es6 thing I've been experimenting with is using async/await, though I try to use it sparingly since it makes the generated javascript harder to follow/debug.

**Comment by vidartf** [comment]
Fixed in #1734
